runtime/caml/mlvalues.h

```diff
@@ -195,14 +195,14 @@ where 0 <= R <= 31 is HEADER_RESERVED_BITS, set with the
 #define Reserved_val(val) (Reserved_hd (Hd_val (val)))
 
 #ifdef ARCH_BIG_ENDIAN
-#define Tag_val(val) (((unsigned char *) (val)) [-1])
+#define Tag_val(val) (((volatile unsigned char *) (val)) [-1])
```

Comment:
A while back I proposed mixed-size atomic accesses to solve issues with races on the header, and the answer was essentially that they are a can of worms memory-model-wise. Now the intent here seems to do a mixed-size relaxed atomic accesses. I am generally in favour of this solution as I did not understand the objections (in particular the Linux kernel does not seem to have problems with mixed-size accesses). But I am curious about  hearing objections people might have. (cc @maranget?)

The alternative is to do a volatile load of the full word followed by truncating the value.

Reply:
Just yesterday, @damiendoligez told us some scary stories about this: in some cases of races involving mixed-size accesses, the resulting behavior is wrong non-memory-model-compliant no matter the number of fences you add, IIUC.

Reply:
I’m afraid I don’t know enough to offer more than this imperfect second-hand account, unfortunately.

Reply:
If the question is given  (1) access the word and mask; or (2) access the byte directly, are (1) and (2)  equivalent in behaviour, I'd venture to say yes, provided all other accesses are to the complete word of to the header byte. Otherwise replacing the complete access by the header byte access would suppress some conflicts of accesses and would potentially introduced relaxed behaviours.  Of course I have no proof...

Mixed sized models for x86 and ARMv8 are studied in [Armed Cats: formal concurrency modelling at Arm](https://inria.hal.science/hal-03470858). Although it is not easy to state that the header access as a byte is ``safe" from the models, I can at least say that (aligned) conflicting writes  (_i.e._ writes that have some bytes in common) are ordered, and that no observation can contradict this ordering. This property is a slight extension of the same basic property applied to non-mixed-size accesses are suggests that nothing nasty can occur when some accesses are replaced by access to the same subpart (here the header).